File: src/kotlin/app/adventofcode/impl/Day03.kt
package adventofcode.impl

import adventofcode.Puzzle
import adventofcode.Helper

class Day03(filename: String) : Puzzle(filename) {

    override fun solveFirst(): String? {
        val diagnostics = input.map { DiagnosticStat(it, it.length, it.bitsToInt()) }

        val commonLength = diagnostics[0].numBits
        if (commonLength == 0 || !diagnostics.all { it.numBits == commonLength }) {
            Helper.debugMsg("Input did not have common length $commonLength")
            return null
        }

        var gammaRate = "";
        var epsilonRate = "";

        for (bitIdx in 0 until commonLength) {
            val (ones, zeros) = countBits(diagnostics, bitIdx)
            gammaRate += (ones > zeros).asChar()
            epsilonRate += (zeros > ones).asChar()
        }

        Helper.debugMsg("Gamma rate was '$gammaRate' or ${gammaRate.bitsToInt()}")
        Helper.debugMsg("Epsilon rate was '$epsilonRate' or ${epsilonRate.bitsToInt()}")

        return "${gammaRate.bitsToInt() * epsilonRate.bitsToInt()}"
    }

    override fun solveSecond(): String? {
        val diagnostics = input.map { DiagnosticStat(it, it.length, it.bitsToInt()) }

        val commonLength = diagnostics[0].numBits
        if (commonLength == 0 || !diagnostics.all { it.numBits == commonLength }) {
            Helper.debugMsg("Input did not have common length $commonLength")
            return null
        }

        val remainingOxygenGenerator = diagnostics.toMutableList()
        val remainingLifeSupport = diagnostics.toMutableList()

        for(bitIdx in 0 until commonLength) {
            if (remainingOxygenGenerator.size > 1) {
                val (ones, zeros) = countBits(remainingOxygenGenerator, bitIdx)
                Helper.debugMsg("Therer are $ones 1s and $zeros 0s remaining in oxygen")
                remainingOxygenGenerator -= { it.statBits[bitIdx] != (ones >= zeros).asChar() }
            }
            if (remainingLifeSupport.size > 1) {
                val (ones, zeros) = countBits(remainingLifeSupport, bitIdx)
                Helper.debugMsg("Therer are $ones 1s and $zeros 0s remaining in lifesupp")
                remainingLifeSupport -= { it.statBits[bitIdx] != (ones < zeros).asChar() }
            }
        }

        if (remainingOxygenGenerator.size == 0 || remainingLifeSupport.size == 0) {
            Helper.debugMsg("There was no valid entries for the bit criteria")
            return null
        }

        val oxygenGeneratorRating = remainingOxygenGenerator[0]
        val lifeSupportRating = remainingLifeSupport[0]

        Helper.debugMsg("Oxygen generator rating was '${oxygenGeneratorRating.statBits}' or ${oxygenGeneratorRating.asNumber}")
        Helper.debugMsg("Life support rating was '${lifeSupportRating.statBits}' or ${lifeSupportRating.asNumber}")

        return "${oxygenGeneratorRating.asNumber * lifeSupportRating.asNumber}"
    }

    fun String.bitsToInt(): Int {
        return this.toInt(2)
    }

    fun countBits(remaining: List<DiagnosticStat>, bitIdx: Int): Pair<Int, Int> {
        val remainingOnes = remaining.filter { it.statBits[bitIdx] == '1' }.size
        return Pair(remainingOnes, remaining.size - remainingOnes)
    }

    fun Boolean.asChar(): Char { 
        return if (this) '1' else '0' 
    }

    operator fun MutableList<DiagnosticStat>.minusAssign(filter: (DiagnosticStat) -> Boolean) { 
        this.removeAll { filter(it) }
    }
}

data class DiagnosticStat(val statBits: String, val numBits: Int, val asNumber: Int)

File: src/csharp/impl/Day03.cs
using System.Text;
using static csharp.Helper;

namespace csharp.impl;

public class Day03 : IPuzzle
{
    private readonly string[] _input;

    public Day03(string filename)
    {
        _input = File.ReadAllLines(filename);
    }

    public string? FirstPuzzle()
    {
        var diagnostics = _input
            .Select(line => new DiagnosticStat(line, line.Length, line.BitsToInt()))
            .ToList();

        var commonLength = diagnostics.FirstOrDefault().NumBits;
        if (commonLength == 0 || !diagnostics.TrueForAll(stat => stat.NumBits == commonLength))
        {
            DebugMsg($"Input did not have common length {commonLength}");
            return null;
        }

        var gammaRateBuilder = new StringBuilder();
        var epsilonRateBuilder = new StringBuilder();

        for (int bitIdx = 0; bitIdx < commonLength; bitIdx++)
        {
            var count = CountBits(diagnostics, bitIdx);
            gammaRateBuilder.Append(BoolToChar(count.ones > count.zeros));
            epsilonRateBuilder.Append(BoolToChar(count.zeros > count.ones));
        }

        var gammaRate = gammaRateBuilder.ToString();
        var epsilonRate = epsilonRateBuilder.ToString();

        DebugMsg($"Gamma rate was '{gammaRate}' or {gammaRate.BitsToInt()}");
        DebugMsg($"Epsilon rate was '{epsilonRate}' or {epsilonRate.BitsToInt()}");

        return (gammaRate.BitsToInt() * epsilonRate.BitsToInt()).ToString();
    }

    public string? SecondPuzzle()
    {
        var diagnostics = _input
            .Select(line => new DiagnosticStat(line, line.Length, line.BitsToInt()))
            .ToList();

        var commonLength = diagnostics.FirstOrDefault().NumBits;
        if (commonLength == 0 || !diagnostics.TrueForAll(stat => stat.NumBits == commonLength))
        {
            DebugMsg($"Input did not have common length {commonLength}");
            return null;
        }

        var remainingOxygenGenerator = diagnostics;
        var remainingLifeSupport = diagnostics;

        for (int bitIdx = 0; bitIdx < commonLength; bitIdx++)
        {
            if (remainingOxygenGenerator.Count > 1)
            {
                var countsOG = CountBits(remainingOxygenGenerator, bitIdx);
                DebugMsg($"Therer are {countsOG.ones} 1s and {countsOG.zeros} 0s remaining in oxygen");

                var criteria = BoolToChar(countsOG.ones >= countsOG.zeros);
                remainingOxygenGenerator =
                    remainingOxygenGenerator.Where(stat => stat.StatBits[bitIdx] == criteria).ToList();
            }

            if (remainingLifeSupport.Count > 1)
            {
                var countsLS = CountBits(remainingLifeSupport, bitIdx);
                DebugMsg($"Therer are {countsLS.ones} 1s and {countsLS.zeros} 0s remaining in lifesupp");

                var criteria = BoolToChar(countsLS.ones < countsLS.zeros);
                remainingLifeSupport = remainingLifeSupport.Where(stat => stat.StatBits[bitIdx] == criteria).ToList();
            }
        }

        if (remainingOxygenGenerator.Count == 0 || remainingLifeSupport.Count == 0)
        {
            DebugMsg($"There was no valid entries for the bit criteria");
            return null;
        }

        var oxygenGeneratorRating = remainingOxygenGenerator.First();
        var lifeSupportRating = remainingLifeSupport.First();

        DebugMsg($"Oxygen generator rating was '{oxygenGeneratorRating.StatBits}' or {oxygenGeneratorRating.AsNumber}");
        DebugMsg($"Life support rating was '{lifeSupportRating.StatBits}' or {lifeSupportRating.AsNumber}");

        return (oxygenGeneratorRating.AsNumber * lifeSupportRating.AsNumber).ToString();
    }

    private static (int ones, int zeros) CountBits(IReadOnlyCollection<DiagnosticStat> remaining, int bitIdx)
    {
        var remainingOnes = remaining.Count(stat => stat.StatBits[bitIdx] == '1');
        return (remainingOnes, remaining.Count - remainingOnes);
    }

    private static char BoolToChar(bool condition) => condition ? '1' : '0';

    private readonly record struct DiagnosticStat(string StatBits, int NumBits, int AsNumber);
}